### ISSUE TITLE
Sorting enums by their values

### ISSUE THREAD

**Issue by PHILLIPS71** [ChilliCream/graphql-platform#7099]
### Product

Hot Chocolate

### Is your feature request related to a problem?

Sorting enums based on their values appears to be unattainable at present. Instead, they are sorted by their names, which may suit certain scenarios but can be cumbersome for others.

### The solution you'd like

Given the example below, I aim to have the statuses sorted based on their values. Hence, when queried, I anticipate seeing entries in the order of Open, Pending, and Closed.

```C#
public enum Status
{
    Open = 0,
    Pending = 1,
    Closed = 2
}
```

Currently, enums are sorted alphabetically by their names. Consequently, the resulting sort order for entries would be Closed, Open, Pending, which is not preferred in this context.

**Comment by glen-84** [comment]
@PHILLIPS71

Hot Chocolate introspection returns the values in declaration order, and BCP uses ordering similar to VS Code (by proximity, etc.):

![image](https://github.com/ChilliCream/graphql-platform/assets/261509/96e30adf-d210-4ce5-a16b-ed6a5d025e14)

Could you please clarify what you meant by "when queried"?

**Comment by PHILLIPS71** [comment]
@glen-84 thanks for looking into this!

I'm referring to if you were to apply a sort order to an enum field it would be nice to have it sort via the value of the enum, instead of the name.

**Comment by glen-84** [comment]
Oh I see.

Could you please provide an example, or at least more details? Is this in-memory, EF Core (which provider), etc.?

**Comment by glen-84** [comment]
1. Which provider are you using?
2. Is the enum mapped to a string type or an enum type in the database?

**Comment by PHILLIPS71** [comment]
> 1. Which provider are you using?
> 2. Is the enum mapped to a string type or an enum type in the database?

1. postgres
2. a string usually, though there are cases an int type is used

**Comment by glen-84** [comment]
Unless I'm missing something, the ordering is simply going to follow the data type of the column – if it's a string, then it'll sort by name, and if it's an integer, then it'll sort by value.

In other words, if you want status to be sorted by value, then consider storing it as an integer.

The only other options that I can think of:

1. Custom SQL that does something like:
    - `ORDER BY CASE status WHEN "Open" THEN 0 WHEN "Pending" THEN 1 WHEN "Closed" THEN 2 END`
        - (possibly not great for performance)
2. Mapping to a PostgreSQL enum, in which case the order is based on the order in which the values are listed when the type is created.

I don't think there's much that we can do here on the Hot Chocolate side.

**Comment by PHILLIPS71** [comment]
I realize now that I've misunderstood the sorting process. I initially thought it would sort based on the enum value after conversion. However, since the conversion occurs only after the result set is returned, the sorting is already completed by that point. I'll close this issue since it was a misunderstanding on my part and use one of the options you've provided. Thanks for clearing this up, I appreciate it!
